_posts/2019-05-10-why-projects-need-codenames.markdown

```diff
@@ -0,0 +1,47 @@
+
+**Branding shifts, products pivot, and companies merge.** Project code names introduce a level of indirection between what engineers build and the labels, URLs, or brands that end users experience. Over a long enough period, these external or superficial changes _will_ happen and risk introducing confusion or just subtle misalignment between form and function. As with software design in general, abstractions can be a powerful tool to [separate concerns](https://en.wikipedia.org/wiki/Separation_of_concerns). Ask yourself: what code changes might be required if your company were to spin off your product or be merged into an acquirer?
+
+**Language is important**, and not just for communication but for shaping our thoughts and assumptions. We frequently find ourselves debating which system should serve as the authority for a given domain model or where to implement a new feature, and project names that overlap with these topics (e.g., `search`, `images`, `suggestions`) would predispose us to certain decisions. Code names free us to focus on the architectural and organizational merits instead.
+
+With this in mind, how should you choose a naming scheme?
+
+### Rules for a project naming scheme
+
+* Avoid implicit value judgments like "new," "next," or "modern." We've all witnessed today's hot project become next year's unloved albatross.
+* Choose a code name scheme that isn't directly related to your technology or business. A flower business using flower names is cute, but breaks down when you want to build a feature that _actually_ is about tulips.
+* Everyone should be able to participate, so avoid industry lingo or anything obscure. Funny story: I work at an art start-up but don't know enough artists to name my projects that way. I struggle even more to pronounce the few projects that are.
+* There should be lots of choices. You'll regret choosing to name your projects after "decathlon sports" or "days of the week" when your team inevitably transitions to microservices.
+* Names should be unique within a company and--ideally--beyond the company's scope. You wouldn't want your `marketing-site` repository to conflict with a contributor's `marketing-site` repository from another organization. Github projects, S3 buckets, Heroku application names, and published libraries all benefit from being globally unique.
+* Bonus points for choosing a rich enough scheme that names can be subtly related to each system's function. At Artsy, we tend to name projects with physics terms, so I'd code-name a shipping service _Weight_.
```

Comment:
I like this, and I think we could expand this point slightly; [Exchange](https://github.com/artsy/exchange) is our order-management repo, which is very apropos, and we can link to it from this post because it's open source.

Reply:
Now that I've made this change and re-read, I realize that "exchange" doesn't exactly sound like a physics term 🤔.